### ISSUE TITLE
Drop support for non-standard combinator arrangements

### ISSUE THREAD

**Issue by nex3** [sass/sass#3340]
* [x] [Proposal](https://github.com/sass/sass/blob/main/accepted/bogus-combinators.md) (https://github.com/sass/sass/pull/3342)

**Deprecation**

* [x] [Tests](https://github.com/sass/sass-spec/issues/1797) (https://github.com/sass/sass-spec/pull/1803)
* [x] [Dart Sass](https://github.com/sass/dart-sass/issues/1727) (https://github.com/sass/dart-sass/pull/1740)
* [x] [Documentation](https://github.com/sass/sass-site/issues/647) (https://github.com/sass/sass-site/pull/654)
* [x] Update canonical spec (https://github.com/sass/sass/pull/3362)

**Removal**

* [ ] [Tests](https://github.com/sass/sass-spec/issues/1797)
* [ ] [Dart Sass](https://github.com/sass/dart-sass/issues/1727)
* [ ] [Documentation](https://github.com/sass/sass-site/issues/647)

---

Currently, Sass is *very* liberal when it comes to where explicit selector combinators can be written—much more so than CSS itself. We allow:

* Multiple combinators between compound selectors (`a > + b`).
* Combinators at the beginning of selectors (`> a`).
* Combinators at the end of selectors (`a >`).

The latter two are useful when nesting style rules, but the former has no known use at all. Historically, we've said that we support these because they enable some browser hacks, but this seems to be purely hypothetical: [browserhacks.com] is the most comprehensive source of hacks I've found, and it doesn't list any hacks that are enabled by this support, even among its "legacy hacks".

[browserhacks.com]: http://browserhacks.com

In addition, supporting these selectors imposes substantial complexity on Sass implementations. They cause bugs (https://github.com/sass/dart-sass/issues/1053) and make the data model more complex than it needs to be. This in turn makes bugs like https://github.com/sass/sass/issues/1807 that don't involve non-standard combinators more difficult to resolve.

Fortunately, these bogus combinators are unlikely to be widely-used in practice and have no effect in real-world browsers, which gives us considerable freedom to drop support for them without needing to do a lengthy deprecation period. We'll need to be careful to preserve the useful aspects of their behavior, and we'll probably want to make these produce errors eventually, but for the time being we can just cause them to omit selectors. In particular:

* If any of a style rule's selectors contains a leading, trailing, or multiple combinator after nesting is resolved, omit that style rule from the generated CSS and emit a deprecation warning.

* If a selector with a leading or trailing combinator is used with any extend-related infrastructure, emit a deprecation warning but proceed as usual for now. Although this isn't really *desirable*, it can produce valid CSS, as in `selector.extend(".a .b", ".a", ".x >")`.

* If a selector with a doubled combinator is used with any extend-related infrastructure, emit a deprecation warning and treat that selector as though it matches no elements (which is in fact the case).

Then, when we release the next major version of Dart Sass, we can start producing errors for the situations we had previously been producing deprecation warnings.

**Comment by nex3** [comment]
I've opened a proposal for this: https://github.com/sass/sass/pull/3342

**Comment by nex3** [comment]
The proposal has landed. Since this almost exclusively affects stylesheets that generate invalid CSS anyway, I don't think it's likely to be controversial so I'm going to give it two weeks for public comment before moving forward with the implementation.

**Comment by kerryj89** [comment]
Will this change make `_header.scss` invalid?  I use this form and it makes for what I feel is a clean separation.

```scss
// layout/_index.scss
.layout-x {
    @include meta.load-css('header');
    ...
}
```

```scss
// layout/_header.scss
> header {
    background: blue;
}
```

**Comment by nex3** [comment]
@kerryj89 That will specifically be invalid because `meta.load-css()` compiles the module's stylesheet to CSS _before_ injecting it into the enclosing context, and `> header` isn't valid CSS. Since `_header.scss` isn't a valid CSS file anyway, I recommend putting it into an explicit mixin, which will continue to work.

**Comment by jdelStrother** [comment]
From https://browserstrangeness.bitbucket.io/css_hacks.html#safari, I was hoping to use this to show a message to Safari 11 users - 

```sass
html >> * #fallback-error { display: block; }
```

but this is now stripped out. Is there any workaround or do I need to find another approach?

**Comment by nex3** [comment]
Since Safari 11.0 is used by substantially fewer users than the 2% cutoff we use for breaking browser changes, I think we'll have to consider it an acceptable loss here. Sorry for the trouble.

**Comment by kerryj89** [comment]
> @kerryj89 That will specifically be invalid because `meta.load-css()` compiles the module's stylesheet to CSS _before_ injecting it into the enclosing context, and `> header` isn't valid CSS. Since `_header.scss` isn't a valid CSS file anyway, I recommend putting it into an explicit mixin, which will continue to work.

Just to confirm I'm understanding you, does this seem okay for [my situation](https://github.com/sass/sass/issues/3340#issuecomment-1174567226)?

```scss
// layout/_index.scss
@use 'header';

.layout-x {
    @include header.content;
    ...
}
```

```scss
// layout/_header.scss
@mixin content {
    > header {
        background: blue;
    }
}
```

**Comment by nex3** [comment]
Yes, that will be valid. You can tell because it doesn't produce a deprecation warning in Sass today.

**Comment by GregJArnold** [comment]
I think I've hit an edge case on this.
```sass
.inputHeader {
  flex-direction: column;
  display: flex;
  
  @include tablet-and-desktop {
    flex-direction: row;
  
    > &:global(.space) {
      margin-top: 0;
    }
  }
} 

@mixin tablet-and-desktop {
  @media (min-width: #{$tablet-min}) {
    @content;
  }
}
```
This seems like it should fall under the nesting exception, because the `@media` selector goes as the outermost block.

**Comment by nex3** [comment]
The issue with that code isn't the media query, it's the fact that the explicit parent selector in `> &:global(.space)` will cause the combinator `>` to appear at the beginning of the compiled selector.

**Comment by javierjulio** [comment]
I'm likely just misunderstanding the [Breaking Change: Invalid Combinators post](https://sass-lang.com/documentation/breaking-changes/bogus-combinators/) but I was worried that it may not be emitting deprecations for CSS I thought would have been invalid in that case. In [this SASS playground snippet, only the first selector emits a deprecation](https://sass-lang.com/playground/#eJyVUk2PgkAMvfMrevLiQnST3QMm/hUyHwUby9QMg9Fs/O874KDo6mE5kFBeX9977XpdBKUZqyDCXdVh06ILaCsjLnhh+MkALHUHVucSyDE5zDWL2W+y+GcJixEB0CrfkMsZ61DCenU4bWL5kl2yrKiJA/rK7NDsKy0n7MaewuyETPoA0OIt+hKcOBy5AbbASiMnwExHEjDULrcx2qOyxvetfhR9Vxv5KJG9tzSYYio11uLxNvqgrCXXlLCCr6u3+ejkMHa0VU3IduxLkWgJQdoYyufQGOuLCW9VwMor10xzyB368NZuKi+fYE/Zf0/yrgLnMlsVUU4dqVGBxD3GNJ5BbpD5P1ltQf3V+4i7a4ivftomUxfyLpwZn1dON8I5+evtJ9bhWRSm9z4eL6iPe3+5kyP6GYtWZt946Z3NjbDEg/NoXwb2C8bM4RE=) but why don't the others do as well? Should they be emitting deprecations? Should I be prefixing those other selectors with `& ...` to avoid any potential breakage in the future?

**Comment by nex3** [comment]
@javierjulio It's still valid to use a leading or trailing combinator when nesting style rules, as long as the final selector that's generated doesn't have them. The reason `+ &` doesn't work on line 4 of your example is because explicitly using `&` disables the normal nesting behavior, so the final selector is `+ .table_tools_segmented_control`. I think what you want is to write `& + &`, which doesn't produce the deprecation warning and generates `.table_tools_segmented_control + .table_tools_segmented_control`.

**Comment by javierjulio** [comment]
@nex3 thank you! The clarification: "explicitly using `&` disables the normal nesting behavior" helped me understand the difference and what I was missing. Yes, I knew to prefix the selector with `& ...` to resolve the deprecation, I just didn't understand why that selector was the only one emitting the deprecation. Thank you again for clarifying!
